### ISSUE TITLE
Intermittent failure of mmap_shared_write_exec_race-32-no-syscallbuf

### ISSUE THREAD

**Issue by rocallahan** [mozilla/rr#2572]
This test was added in https://github.com/mozilla/rr/commit/6f414dd18a76b49cb571a32b4312a9c4613a83ad but the fix is incomplete.

The problem is here:
```
            for (auto tt : v->task_set()) {
              // If the task here has execed, we may not be able to record its
              // memory any longer, so loop through all tasks in this address
              // space in turn in case any *didn't* exec.
              if (static_cast<RecordTask*>(tt)->record_remote_fallible(km.intersect(mr)) > 0) {
                break;
              }
            }
```

These `record_remote_fallible` calls can be racing with an `execve` in the task. If the `execve` switches address spaces just before the `record_remote_fallible`, *and* the post-exec address space has memory at the requested address, we'll assign those memory values to the corresponding address in the old address space, which is wrong.

**Comment by rocallahan** [comment]
Clearly it's more likely that the post-exec address space will have memory at the requested address for 32-bit processes, but it could happen for 64-bit too of course.

Unfortunately it's not obvious how to fix this. Reading through any task is unsafe if there's any other thread in the same thread-group doing an execve. Even with only one task in the thread-group, if we decide not to record memory because it's in an execve, but the execve fails, then we haven't updated that task's memory correctly.

**Comment by rocallahan** [comment]
A thought: we made execve `ALLOW_SWITCH` because we needed the scheduler to allow other tasks to run, so we could observe unstable exits of threads in the same thread-group, which must be allowed to proceed past `PTRACE_EVENT_EXIT` before the execve returns. But *maybe* we could leverage the logic Keno added in 4e10088344fba26eac97c6d76836d1baaba96cdd to allow us to synchronously handle all task exits without scheduling any other tasks to run, so effectively making execve PREVENT_SWITCH. This would simplify things quite a bit.

**Comment by rocallahan** [comment]
How nice it would be if we could wait on a selected set of tasks, or at least for all tasks in a specific threadgroup even though we're ptracing them all. I guess with pidfds we could do that, but I don't think we should depend on pidfds yet.

**Comment by rocallahan** [comment]
How about: when a task is in an `execve` syscall, `Scheduler::reschedule` will not schedule any task, i.e. treat all tasks as blocked. Then it will wait for state changes in any task, which will catch execve completing (successfully or not), or tasks exiting, or other task syscalls completing. If we see a state change in a thread group that's not involved in the execve, just ignore it (like unlimited_ticks_mode does) and keep waiting.

**Comment by rocallahan** [comment]
@Keno any thoughts on the above?

**Comment by Keno** [comment]
I think it should be possible to shepherd a task group through execve without ever going back to the scheduler. In theory we can just waitid with the thread group I'd and pick off the exit events one by one.

**Comment by rocallahan** [comment]
Does `waitid` with `P_PID` actually wait for any task in the threadgroup, even when the the specified id is ptraced? (I.e. NOT like `waitpid` in this case?)

**Comment by Keno** [comment]
Oh, no you're right. I didn't realize that there was no P_TGID. All the wait machinery in the kernel supports it, but it's not exposed to userspace.

**Comment by rocallahan** [comment]
OK, so any thoughts on my proposal?

**Comment by Keno** [comment]
Yes, I think just waiting for everything ignoring events in other thread groups would be ok.

**Comment by rocallahan** [comment]
I discovered another way racing with `execve` can break things: it causes CLOEXEC fds to close asynchronously. So, for example, we may have an fd with a `FileMonitor` attached that is CLOEXEC, and the `execve` races with another task with the same fd table opening a syscallbuf desched fd. The `execve` closes the fd in the kernel, but before we know about that we see the fd associated with a new `PreserveFileMonitor` while it already had a (possibly different) monitor.

**Comment by rocallahan** [comment]
>  it causes CLOEXEC fds to close asynchronously

Actually I think this is not true, the fd effects of the execve should only be observable inside the execve task itself, since post-execve it doesn't share the fd table with any other task.

**Comment by rocallahan** [comment]
ee6df92190c05999e86c75acf8a788a9e4839bc5 hopefully has fixed this.
